pkg/cluster-runtime/apps.go

```diff
@@ -98,38 +96,39 @@ func (i AppInstaller) LaunchClusterImage(master0 net.IP, launchCmds []string) er
 			return fmt.Errorf("failed to render launch cmd: %v", err)
 		}
 
-		if err = i.infraDriver.CmdAsync(master0, fmt.Sprintf(common.CdAndExecCmd, clusterRootfs, cmdline)); err != nil {
+		if err = i.infraDriver.CmdAsync(master0, fmt.Sprintf(common.CdAndExecCmd, appPath, cmdline)); err != nil {
 			return err
 		}
 	}
 
-	return i.save()
+	return i.save(common.GetDefaultApplicationFile())
 }
 
 // todo save image info to disk or api server, we need new interface to do this.
-func (i AppInstaller) save() error {
-	var extensionList []v12.ImageExtension
-	applicationFile := common.GetDefaultApplicationFile()
+func (i AppInstaller) save(applicationFile string) error {
+	f, err := os.OpenFile(applicationFile, os.O_RDWR|os.O_APPEND|os.O_CREATE, 0600)
+	defer func() {
+		_ = f.Close()
+	}()
 
-	if osutils.IsFileExist(applicationFile) {
-		b, err := ioutil.ReadFile(filepath.Clean(applicationFile))
-		if err != nil {
-			return err
-		}
+	if err != nil {
+		return err
+	}
 
-		b = bytes.TrimSpace(b)
-		if len(b) != 0 {
-			if err := json.Unmarshal(b, &extensionList); err != nil {
-				return fmt.Errorf("failed to load default application file %s: %v", applicationFile, err)
-			}
-		}
+	err = syscall.Flock(int(f.Fd()), syscall.LOCK_EX|syscall.LOCK_NB)
+	if err != nil {
+		return fmt.Errorf("cannot flock file %s - %s", applicationFile, err)
 	}
 
-	extensionList = append(extensionList, i.extension)
-	content, err := json.Marshal(extensionList)
+	content, err := json.Marshal(i.extension)
 	if err != nil {
 		return fmt.Errorf("failed to marshal image extension: %v", err)
 	}
 
-	return osutils.NewCommonWriter(applicationFile).WriteFile(content)
+	_, err = f.Write(content)
+	if err != nil {
```

Comment:
unify use 
if err:=func();err!=nil{}

please check the whole picture for yourself